service/agentconf.go

```diff
@@ -192,23 +192,34 @@ func (s *systemdServiceManager) WriteSystemdAgents(
 		if !ok {
 			return nil, nil, nil, errors.Errorf("%s service not of type UpgradableService", svcName)
 		}
+
+		dbusMethodNotFound := false
 		if err = uSvc.WriteService(); err != nil {
-			logger.Infof("failed to write service for %s: %s", agentName, err)
-			errAgentNames = append(errAgentNames, agentName)
-			lastError = err
-			continue
+			// Note that this error is already logged by the systemd package.
+
+			// This is not ideal, but it is possible on an Upstart-based OS
+			// (such as Trusty) for run/systemd/system to exist, which is used
+			// for detection of systemd as the running init system.
+			// If this happens, then D-Bus will error with the message below.
+			// We need to detect this condition and fall through to linking the
+			// service files manually.
+			if strings.Contains(err.Error(), "No such method 'LinkUnitFiles'") {
```

Comment:
At the very least, wouldn't we want to log the original error in the following Infof?
Also, do we know if this message is reasonably stable?
I assume there isn't an error code or any other clearer message than this that is *intentionally* giving the message, rather than accidentally.

I guess if we are only really focused on Trusty, whatever we have there isn't really going to be changing.

Reply:
I agree; matching error strings smells. I looked through https://github.com/coreos/go-systemd and there is no typed error or code, for this.

Logged output looks like this:
```
machine-0: 10:37:31 ERROR juju.service.systemd dbus link request failed for application "jujud-machine-0": No such method 'LinkUnitFiles'
machine-0: 10:37:31 DEBUG juju.service.systemd stack trace:
No such method 'LinkUnitFiles'
/home/joseph/go/src/github.com/juju/juju/service/systemd/service.go:595: dbus link request failed for application "jujud-machine-0"
machine-0: 10:37:31 INFO juju.service attempting to manually link service file for machine-0
```

Reply:
I've made the matching a little softer, but I don't think there is much more we can do with it...